### PR TITLE
feat: add initial healthcheck endpoint (ampd)

### LINKED ISSUE #12: Create an HTTP `/status` endpoint for healthcheck probing (ampd)
### Context

Some runtimes like Kubernetes need a way to understand if the service it's healthy or not. 

### Problem

`ampd` does currently not have a healthcheck.

### Goals

Study and implement the simplest HTTP `/status` endpoint. It should return a `200 OK` if the service is healthy or an `5xx` response if not, plus some info regarding the problem[s] if we are able to do it.

### PR DESCRIPTION
**Pull request by eloylp**
## Description

At [Eiger](https://github.com/eigerco) we are working on the Axelar Solana implementation.  During this work, we realised `ampd` doesn't have a defined way to check the `liveness` of the process from an external perspective. This is specially useful when deploying workloads in i.e Kubernetes, as the cluster can query the process in order to check if its alive or not.

This is a PR proposal for adding an `HTTP` `/status` endpoint to the `ampd` process, that can be used by external systems to determine the `liveness` of the process.

Closes #12

## Todos

- [x] Unit tests
- [x] Manual tests See this [comment](https://github.com/eigerco/axelar-amplifier/pull/14#issuecomment-2047856562) .
- [x] Documentation
- [x] Connect epics/issues

### Manual
1. Start the `ampd` server.
2. ```bash
    $ curl localhost:3000/status
    ```
### Unit tests

```bash
$ cd ampd && cargo test
```

## Expected Behaviour

Once `ampd` starts its activity, an HTTP request to the `/status` endpoint should return a `200 OK` response in json format:

```bash
{
 ok: true
}
```

## Other Notes

This is just a dummy HTTP health check endpoint that only provides a probe of liveness of the process. It doesn't query any subsystem in order to check their internal health. In a future, such option could be revisited.

### CODE REVIEW

File: ampd/src/health_check.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+impl Server {
+    pub async fn new(bind_addr: SocketAddr) -> Result<Self, HealthCheckError> {
+        Ok(Self {
+            listener: tokio::net::TcpListener::bind(bind_addr)
+                .await
+                .change_context(HealthCheckError::Error(format!(
+                    "Failed binding to addr: {}",
+                    bind_addr
+                )))?,
```
Comment (eloylp) [inline ampd/src/health_check.rs:27]: Normally i do not like to use `new()` for doing things like binding sockets. Initially i consider this approach doable taking into account how the rest of the codebase interacts with this server.

File: ampd/src/health_check.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    #[allow(dead_code)]
+    pub fn listening_addr(&self) -> Result<SocketAddr, HealthCheckError> {
+        Ok(self
+            .listener
+            .local_addr()
+            .map_err(|e| HealthCheckError::Error(e.to_string()))?)
+    }
+
+    pub async fn run(self, cancel: CancellationToken) -> Result<(), HealthCheckError> {
+        tracing_subscriber::fmt::init();
```
Comment (roberts-ivanovs) [inline ampd/src/health_check.rs:40]: The `main.rs` file already initialises the tracing subscriber. afaik initialising it the second time will result in a panic

**Review by roberts-ivanovs — APPROVED** [review]

File: ampd/src/health_check.rs (hunk 2 above)
Comment (eloylp) [reply ampd/src/health_check.rs]: Thanks ! i was missing to cleanup that part. Fixed tracing at 9937cd5efe25c104a8c9e24b4cd46429b69d6c40

**Comment by eloylp** [comment]
### Manual testing

In order to provide certain confidence the healthcheck endpoint is not disturbing other components, the `ampd` daemon was tested by:

1. Bringing up locally the [tofnd](https://github.com/axelarnetwork/tofnd) (pre-requisite). We used the docker-compose way sharing the port with the host.

2. Configure the `~/.ampd/config.toml` file with the following content:

   ```toml
   tm_jsonrpc = "https://axelar-testnet-rpc.qubelabs.io:443"
   tm_grpc = "http://axelar-testnet-grpc.qubelabs.io:9090"
   event_buffer_capacity = 1000
   health_check_bind_addr = "0.0.0.0:3000"
   
   [service_registry]
   cosmwasm_contract="axelar1hrv2zf8xsfsey2umewng68c8pwsyktq4sm8u6k5sch4gqkn0qxrsg28xnx"
   
   [broadcast]
   batch_gas_limit="10000000"
   broadcast_interval="1s"
   chain_id="devnet-amplifier"
   gas_adjustment="2"
   gas_price="0.00005uamplifier"
   queue_cap="1000"
   tx_fetch_interval="600ms"
   tx_fetch_max_retries="10"
   
   [tofnd_config]
   key_uid="axelar"
   party_uid="ampd"
   url="http://127.0.0.1:50051"
   
   [[handlers]]
   type = 'MultisigSigner'
   cosmwasm_contract = 'axelar1lz2hevr93dwa3l86n7hnyy36age39jsk0e8nsj804xr3n8hh958q79m2sq'
   ```

3. Execute `cd ampd && cargo run` :

   ```bash
   ❯ cargo run
       Finished dev [unoptimized + debuginfo] target(s) in 1.19s
        Running `/home/eloylp/projects/eiger/axelar-amplifier/target/debug/ampd`
   2024-04-10T13:15:35.305195Z  INFO ampd: found config file /home/eloylp/.ampd/config.toml
   2024-04-10T13:15:35.308031Z  INFO ampd: starting daemon args=Args { config: ["~/.ampd/config.toml", "config.toml"], state: "~/.ampd/state.json", output: Output::Text, cmd: () }
   2024-04-10T13:15:35.308152Z  INFO ampd::state: loading state from disk
   2024-04-10T13:15:37.850389Z  INFO ampd::health_check: Starting health check server at: 0.0.0.0:3000
   2024-04-10T13:15:40.417094Z  INFO ampd::state: state updated handler="multisig-signer" height=12887365
   ```

   We can see the output `Starting health check server at: 0.0.0.0:3000` , which announces the healthcheck endpoint should be reachable at this point.

4. Reach the health endpoint:

   ```bash
   ❯ curl localhost:3000/status
   {"ok":true}
   ```

**Pull request closed without merging**